p2p/bucket/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+//! Bucket data structure
+//!
+//! A collection data structure that discriminates its unique items and place them into "buckets"
+//!
+//! The item must implement the [`Bucketable`] trait that defines how to create the discriminant
+//! from the item type. The data structure will internally contain any item into "buckets" or vectors
+//! of sized capacity N that regroup all the stored items with this specific discriminant.
+//!
+//! A practical example of this data structure is for storing N amount of IP discriminated by their subnets.
+//! You can store in each "buckets" corresponding to a /16 subnet up to N IPs of that subnet.
+//!
+
+use arrayvec::ArrayVec;
+use std::{collections::BTreeMap, net::Ipv4Addr};
+
+/// A discriminant can be compute from the type
+pub trait Bucketable: Sized + Eq + Clone {
+    /// The type of the discriminant being used in the Binary tree.
+    type Discriminant: Ord + AsRef<[u8]>;
+
+    /// Method that can compute the discriminant from the item.
+    fn discriminant(&self) -> Self::Discriminant;
+}
+
+/// A collection data structure discriminating its unique items
+/// with a specified method. Enable a fair distribution of the
+/// items based on their discriminants when popped.
+pub struct Bucket<const N: usize, I: Bucketable> {
+    /// The storage of the bucket
+    storage: BTreeMap<I::Discriminant, ArrayVec<I, N>>,
+    /// Internal round-robin state
+    round_robin: usize,
```

Comment:
for now can you remove this round_robin pop impl please.

It should be done as a separate struct which impls `Iterator`, however until we need it I would just remove this.

Reply:
I'm actually planning on passing to the Address book PR after that. Shouldn't it be preferable to make that iterator structure in this PR instead ?

Reply:
The bucket structure is more for making sure we store a wide range of IPs, not that we select a wide range of IPs, having this round_robin makes it more likely for consecutive subnets to be all the peers we connect to.

Just randomly selecting peers across the whole structure would be better IMO, but this can be considered later.

Reply:
Fair enough.

Reply:
If I remove the current pop there is no method to actually extract an IP address from the storage. Do you want me to make the `storage` field public ? or adapt the current `pop` to take a random item (i know this contradict your statement of needing an iterator but I'm going to implement anyway in the next PR since well I'll need it)?

Reply:
Yeah adding a `pop` method that selects  a random peer would be good - although it should leave the peer in structure and just return a reference